snuba/clickhouse/native.py

```diff
@@ -127,6 +131,67 @@ def close(self) -> None:
             pass
 
 
+T = TypeVar("T")
+R = TypeVar("R")
+
+
+def transform_nullable(
+    function: Callable[[T], R]
+) -> Callable[[Optional[T]], Optional[R]]:
+    def transform_column(value: Optional[T]) -> Optional[R]:
+        if value is None:
+            return value
+        else:
+            return function(value)
+
+    return transform_column
+
+
+def transform_date(value: date) -> str:
+    """
+    Convert a timezone-naive date object into an ISO 8601 formatted date and
+    time string respresentation.
+    """
+    # XXX: If the original value had a valid nonzero UTC offset, this will
+    # result in an incorrect value being returned.
+    return datetime(*value.timetuple()[:6]).replace(tzinfo=tz.tzutc()).isoformat()
+
+
+def transform_datetime(value: datetime) -> str:
+    """
+    Convert a timezone-naive datetime object into an ISO 8601 formatted date
+    and time string representation.
+    """
+    # XXX: If the original value had a valid nonzero UTC offset, this will
+    # result in an incorrect value being returned.
+    return value.replace(tzinfo=tz.tzutc()).isoformat()
+
+
+def transform_uuid(value: UUID) -> str:
+    """
+    Convert a UUID object into a string representation.
+    """
+    return str(value)
+
+
+transform_column_types = build_result_transformer(
+    [
+        (re.compile(r"^Date(\(.+\))?$"), transform_date),
+        (
+            re.compile(r"^Nullable\(Date(\(.+\))?\)$"),
+            transform_nullable(transform_date),
+        ),
+        (re.compile(r"^DateTime(\(.+\))?$"), transform_datetime),
+        (
+            re.compile(r"^Nullable\(DateTime(\(.+\))?\)$"),
+            transform_nullable(transform_datetime),
+        ),
+        (re.compile(r"^UUID$"), transform_uuid),
+        (re.compile(r"^Nullable\(UUID\)$"), transform_nullable(transform_uuid)),
```

Comment:
This is kind of verbose and I think we'll eventually want to clean this up so that each type doesn't need to have an explicit `Nullable(…)` wrapper. I can't think of a clean way to do that right now without hiding that the first tuple member needs to be valid regular expression, and would need to be able to be formatted into `Nullable(%s)` without problems. (I don't think passing a string here and creating a two patterns out of it later is a very good option.)

Reply:
I guess another option here that I just thought of would be to match on the `Nullable\(.+\)` pattern first to unwrap the inner type, wrapping the column transformation function in `transform_nullable` dynamically. Probably a bit faster too as it'd reduce the space of this linear scan when trying to find the right column transformer.

Reply:
Made that change in 527b872.